docs: add assertions docs (#506)
Redocly/openapi-cli: docs/resources/rules/assertions.md

## Assertion object

Property | Type | Description
-- | -- | --
on | `string` \| [`string`] | **REQUIRED.** The node type, or list of node types, to the lint targets. Nodes are traversed with a dot `.` between nodes, and can start from any [OpenAPI node type](#openapi-node-types). A special value `$keys` can be used to target the keys of any map (for example, `ResponseMap.$keys`).
message | `string` | Problem message displayed if the assertion is false.
suggest | [`string`] | List of suggestions to display if the problem occurs.
severity | `string` | The severity level of the problem if the assertion is false. It must be one of these values: `error`, `warn`, `off`. Default value is `error`.
enum | [`string`] | Asserts value is within a predefined list of values. See [enum example](#enum-example).

> pattern | `string` | Asserts value matches a regex pattern. See [regex pattern example](#regex-pattern-example).

Reviewer: ```suggestion
pattern | `string` | Asserts value matches a regex pattern. See [regex pattern example](#pattern-example).
```

casing | `string` | Asserts a casing style from this possible list: `camelCase`, `kebab-case`, `snake_case`, `PascalCase`, `MACRO_CASE`, `COBOL-CASE`, `flatcase`. See [casing example](#casing-example).

> mutuallyExclusive | [`string`] | Asserts list of properties (key names only) are mutually exclusive. Use `$keys` in the `on` when this is enabled (for example, `Operation.$keys`). See [mutuallyExclusive example](#mutuallyexclusie-example).

Reviewer: ```suggestion
mutuallyExclusive | [`string`] | Asserts list of properties (key names only) are mutually exclusive. Use `$keys` in the `on` when this is enabled (for example, `Operation.$keys`). See [mutuallyExclusive example](#mutuallyexclusive-example).
```

property | `string` | **REQUIRED.** The name of the property in the collection of objects that will hold values to assess the sort order.

## Examples

The following example shows two assertions each multiple asserts (`defined`, `minLength`, `maxLength`, `pattern`).

The `Operation`, `Tag`, and `Info` properties must:
- be defined
- have at least 30 characters
- end with a _full stop_.

In addition, the `Operation` summary property must:
- be defined
- be between 20 and 60 characters
- not end with a _full stop_.

The following shows how to write that configuration:

```yaml
lint:
  rules:
    assertions:
      - on:
          - Tag.description
          - Operation.description
          - Info.description
        message: Description must be at least 30 characters and end with a full stop.
        severity: error
        defined: true
        minLength: 30
        pattern: /\.$/
      - on:
          - Operation.summary
        message: Operation summary must be at least 20 characters and not end with a full stop.
        severity: error
        defined: true
        minLength: 20
        maxLength: 60

>         pattern: /^\.$/

Reviewer: ```suggestion
        pattern: /[^\.]$/
```

The following example asserts the operation `description` and `externalDocs` must be mutually exclusive.
This assert runs only on node's keys.
Append `$keys` to the `on` node value to run on the node's keys.

```yaml
lint:
  rules:
    assertions:
      - on: Operation.$keys
        message: "Operation must not define both properties together: description and externalDocs"
        severity: error
        mutuallyExclusive:
          - description
          - externalDocs
```

### `mutuallyRequired` example

The following example asserts that a response body schema must have both `amount` and `currency` properties (and not either one by itself).
This assertion runs only on node's keys.
Append `$keys` to the `on` node value to run on the node's keys.

```yaml Schema example
lint:
  rules:
    assertions:
      # - on: Operation.Response.$keys check this
        message: The created_at and updated_at properties are mutually required
        severity: error
        mutuallyRequired:
          - created_at
          - updated_at
```

The following example asserts that `PUT` requests have both `200` and `201` responses defined.

```yaml Response example
lint:
  rules:
    assertions:

>       - on: PathItem.put.responses.$keys

Reviewer: We do not support properties inside properties, only property inside NodeType.

        undefined: true
```

### `nonEmpty` example

The following example asserts that the operation summary is not empty.

```yaml
lint:
  rules:
    assertions:
      - on: Operation.summary
        message: Operation summary should not be empty
        severity: error
        nonEmpty: true
```

### `minLength` example

The following example asserts that the minimum length of each operation summary is 20 characters.

```yaml
lint:
  rules:
    assertions:
      - on: Operation.summary
        message: Operation summary must have minimum of 20 chars length
        severity: error
        minLength: 20
```

### `maxLength` example

The following example asserts that the maximum length of each operation summary is 20 characters.

```yaml
lint:
  rules:
    assertions:
      - on: Operation.summary
        message: Operation summary must have maximum of 20 characters
        severity: error
        maxLength: 20
```

### `sortOrder` example

Sort order can be applied on lists (arrays) of strings or lists of objects (collections).

The following example asserts that the status schema's enum list is in alphabetical order.

```yaml
lint:
  rules:
    assertions:

>       - on: Schema.status.enum

Reviewer: We do not support properties inside properties, only property inside NodeType.

        message: The status enums should be in alphabetical order
        suggest:
          - alphabetize the list of status enums
        severity: error
        sortOrder: asc
```

The following example asserts the tag names (tags are a collection) are sorted in alphabetical order.

The `property` field is **REQUIRED** to sort collections based on that property's value. 

```yaml
lint:
  rules:
    assertions:
      - on: Operation.tags
        message: Tags should be ordered by name in alphabetical order
        severity: error
        sortOrder:
          direction: asc
          property: name
```

## OpenAPI node types

Redocly defines a type tree based on the document type.
OpenAPI 2.0 has one type tree.
OpenAPI 3.0 and OpenAPI 3.1 share a type tree.

Node types are traversed using dot notation.

The following example shows ways to traverse to the equivalent destination.
One starts from the definition root into the info description.
The other starts from the info section.

```yaml Traversal from DefinitionRoot object type
lint:
  rules:
    assertions:
      - on:

>         - DefinitionRoot.info.description

Reviewer: ```suggestion
        - DefinitionRoot.Info.description
```

Author: I don't understand. Can you traverse objects but not regular properties?

Reviewer 2: No, this is not about traversal at all. I'll explain you on a call.

minLength | `integer` | Asserts a minimum length (inclusive) of a string or list (array). See [minLength example](#minlength-example).
maxLength | `integer` | Asserts a maximum length (exclusive) of a string or list (array). See [maxLength example](#maxlength-example).
sortOrder | `string` \| [Sort Order object](#sort-order-object) | Asserts a sort order to a list (array) of strings or a list of objects by one of the object's property values. See [sortOrder example](#sortorder-example).

## Context object

Property | Type | Description
-- | -- | --
type | `string` | **REQUIRED.** One of the [OpenAPI node types](#openapi-node-types).
filterKeys | [`string`] | The list of key names to continue evaluation with respect to the subject.

See the [context example](#context-example).

## Sort Order object

Property | Type | Description
-- | -- | --
direction | `string` | **REQUIRED.** Asserts values are sorted in ascending or descending order using one of these possible values: `asc` or `desc`.
property | `string` | **REQUIRED.** The name of the property in the collection of objects that will hold values to assess the sort order.

## Examples

> The following example shows two assertions each multiple asserts (`defined`, `minLength`, `maxLength`, `pattern`).

Reviewer: Frankly I see four but not two assertions in example below, but I understand that technically there are only two :) I think you divided it because of message as it should be exact about the subject. We could combine it within `subject` key, like

```yaml
lint:
  rules:
    assertions:
      - subject: 
        - name: Tag
          message: Tag description must be at least 30 characters and end with a full stop.
        - name: Operation
          message: Operation description must be at least 30 characters and end with a full stop.
        - name: Info
          message: Info description must be at least 30 characters and end with a full stop.
        property: description
        severity: error
        defined: true
        minLength: 30
        pattern: /\.$/
```
If the message is about the property, we can put it under `property` key, for example
```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        property: 
        - name: description
          message: Tag description must be at least 30 characters and end with a full stop.
        - name: name
          message: Tag name must be at least 30 characters and end with a full stop.
        severity: error
        defined: true
        minLength: 10
        pattern: /\.$/
```
If we want to use one message for all the asserts, so let's do it as previous at the assertion level.
So now we will have three ways to use `message` - under `subject`, under `property` and at the assertion level.

Author: Actually, you are right. I adjusted it to four assertions but didn't show it here. We are trying to limit nesting as much as possible. While I am okay with the idea of make subject and property lists, I think it may result in overcomplexity and documenting it would be more difficult than making four assertions.

Author: ```suggestion
The following example shows four assertions each multiple asserts (`defined`, `minLength`, `maxLength`, `pattern`).
```

Author: I think this is okay for properties:  ` string | [string] ` but not something like ` string | [NameMessage Object]`. Flatter with more assertions is easier to use.

The `Operation`, `Tag`, and `Info` properties must:
- be defined
- have at least 30 characters
- end with a _full stop_.

In addition, the `Operation` summary property must:
- be defined
- be between 20 and 60 characters
- not end with a _full stop_.

The following shows how to write that configuration:

```yaml
lint:
  rules:
    assertions:
      - subject: Tag

>         property: description

Reviewer: Agree that the key name should be `properties` and it should be an array, so we can evaluate on more that one property from `subject` Node type. But how about nested properties? Is it possible case when we have prop1.prop2 in some Node type?

Reviewer 2: As Adam mentioned above better to keep it flat to simplify docs/usage.

I'm not sure about nested properties. We can technically support it but I'm afraid people may think it will traverse through $ref's... I would keep it simple first and then make it more complex if needed.

Author: No nesting.  I'm okay to make it properties like: `string` | [`string`] type. I think it could be helpful.

Author: I noticed it was already defined like `string | [string]`. I went to rename it to properties, and then as I was updating the examples I become uncertain if `properties` is a better name than `property` and decided to stick with `property`. I have a small concern that `properties` is overloaded and may be confusing in the documentation between OpenAPI/JSON Schema and that should dominate compared to our "assertions".

        defined: true
        minLength: 30
        pattern: /\.$/
      - subject: Info
        property: description
        message: Info description must be at least 30 characters and end with a full stop.
        severity: error
        defined: true
        minLength: 30
        pattern: /\.$/
      - subject: Operation
        property: summary
        message: Operation summary must be between 20 and 60 characters and not end with a full stop.
        severity: error
        defined: true
        minLength: 20
        maxLength: 60
        pattern: /[^\.]$/
```

### `context` example

The following example asserts that PUT responses with HTTP status 200 or 201 cannot return an application/pdf content type.
Without the `context`, the assertion would evaluate every MediaTypeMap including:
- Responses with all codes including others than 200 or 201
- Responses for all HTTP methods including delete, get, post, and more.
To restrict the evaluation, use the context feature to limit where context is used.

```yaml
- context:
    - type: Operation
      filterKeys: [put]
    - type: ResponsesMap
      filterKeys: [201, 200]
  subject: MediaTypeMap
  disallowed: ['application/pdf']

> ```

Reviewer: I like the idea of `context`, especially how it expands the previous "dot notation" way to traverse Node types. But I see there couple things to improve:

1. It's not clear from the notation above what is the Node types hierarchy - I understand that it's Operation -> ResponsesMap, but would be nice to introduce it clearer, like `next` key (or more suitable word) 
    ```js
    - context:
        - type: Operation
          filterKeys: [put]
          next:
          - type: ResponsesMap
            filterKeys: [201, 200]
    ```
    It will also expand the `context` power - now `context` will spread in many directions, like in the next example 
    ```js
    - context:
        - type: Operation
          filterKeys: [put]
          next:
          - type: ResponsesMap
            filterKeys: [201, 200]
        - type: Type1
          filterKeys: [key1, key2]
          next:
          - type: Type2
            filterKeys: [key3]
            next:
            - type: Type3
              filterKeys: [key4]     
    ```
    It tells us that the first context "direction" is Operation -> ResponsesMap, and the second Type1 -> Type2 -> Type3

2. I would rather use `onKeys` not `filterKeys`, as it sounds like "filter out". For me it's clearer :) 

3. It would be nice to have ability to exclude keys, with `offKeys` (or more suitable word) :
    ``` js
    - context:
      - type: Operation
        onKeys: [put]
        next:
        - type: ResponsesMap
          offKeys: [201, 200]
    ```

Reviewer 2: @ohorbachevskyi thanks for the great feedback!

1. I believe you did not get the context correctly. The context is not an array of alternative contexts, this is actually a hierarchy. I see why you got confused though. But I think your suggestion is over-complicated and won't be used in any real use-cases. I would keep it as is.

2. I believe we were going to rename it to `filterParentKeys`. But I like `onParentKeys` much more. @adamaltman what do you think?

3. I'm not sure if `off` is the correct usage here but I agree it would be nice to have some keyword like that.

Author: How about `matchKeys` and `excludeKeys`?

Reviewer 2: @adamaltman I think it's important to add `parent` word into naming... so `matchParentKeys`, `excludeParentKeys`?

Reviewer: @adamaltman Yeah, `matchKeys` and `excludeKeys` much better! @RomanHotsiy why parent is so important here?

Author: I think it depends on how well people read and understand the docs. The issue is I only understand it personally as the parent. Let's take `put` for example... I think of it as belonging to the `PathItem` but it won't work as a `matchKey` of the `PathItem`. Instead, it is viewed as the parent of the `Operation`.

Reviewer: I got it, thnx for clarifying.

      - subject: Operation
        property: summary
        message: Operation summary must have a maximum of 20 characters
        severity: error
        maxLength: 20
```

### `sortOrder` example

Sort order can be applied on lists (arrays) of strings or lists of objects (collections).

The following example asserts that the status schema's enum list is in alphabetical order.

```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: enum
        context:
          - type: Schema
            filterKeys: 
              - status
        message: The status enums should be in alphabetical order
        suggest:
          - alphabetize the list of status enums
        severity: error
        sortOrder: asc
```

The following example asserts the tag names (tags are a collection) are sorted in alphabetical order.

The `property` field is **REQUIRED** to sort collections based on that property's value. 

```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        property: name
        message: Tags should be ordered by name in alphabetical order
        severity: error
        sortOrder:
          direction: asc
          property: name
```

> <!-- TODO: Does the sortOrder need a property "property" now? -->

Reviewer: Yep, you're right, now a property "property" is redundant here; 
and we can use the simpler `sortOrder` notation like `sortOrder: asc`, because we can set the order's property right after the `subject`.

Author: I've adjusted it (I think)

Author: We need to look at other options. I'm pinging @RomanHotsiy (and discussed it with him in a meeting).

Reviewer 2: I don't like the extra `property` in the `sortOrder`. This will be confusing. We should come up with some better idea.

I think we are mixing a few things here. Sorting by a property value and sorting the property value itself.

I don't think we should rely on the type of the property value to determine the behaviour because some properties can have mixed values, e.g. `type` can be string or array.

One possible solution is to bind the behaviour to our type tree. We clearly know whether some type is a collection or a map (usually these have `_List` or `_Map` suffix). So if the type is a collection we should use `property` as a sort key vs sort the `property` itself:

**Sort collection by property `name`:**
```yaml
lint:
  rules:
    assertions:
      - subject: Tag_List
        property: name
        sortOrder: asc
```

**Sort value of a property:**
```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: enum
        sortOrder: asc
```

The main concern with this approach is we can't actually sort something that is untyped (e.g. vendor extensions). Imagine x-metadata vendor extension which may look like:

```yaml
x-metadata:
  - name: Description
    value: Lorem ipsum
  - name: Location
    value: Lviv
```

```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: x-metadata # how to sort by `name` ?
        sortOrder: asc # what sort order even means here ?
```

Author: Here is another idea. What about two different sort assertions:

- `sortCollection`
- `sortValues`

### Sorts the **collection** by the property `name`
```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        property: name
        sortCollection: asc
```

### Sorts the **values** of the property `enum`
```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: enum
        sortValues: asc
```

For something that may have mixed values, like `product` in this example:

```yaml
x-metadata:
  - name: Sample name as a string
     product: elf
  - name: What is this?
     product:
        - bar
        - foo
```

Someone can specify if they want to sort by the **values** or by the collection by choosing the appropriate assertion `sortValues` or `sortCollection`. 

The open question is what is the logic of `sortCollection` when the value is not a scalar. In the example above, will it use the array's first value for the comparison logic? Or what?

Author: > One possible solution is to bind the behaviour to our type tree. We clearly know whether some type is a collection or a map (usually these have _List or _Map suffix). So if the type is a collection we should use property as a sort key vs sort the property itself:

One thing I don't like about this idea is that we already have types for some maps like `ResponsesMap`, so this feels like we missed types like `TagsMap`. I think we are missing content (tech docs) on the type tree. And this idea may also make things like typeahead less intuitive too.

Reviewer 2: > What about two different sort assertions:

I don't like this idea because `sortValues` and `sortCollection` feel more like a decorator, not as an assertion.

> The open question is what is the logic of sortCollection when the value is not a scalar. 

Also `sortCollection` has other open questions. What if the subject is not part of any collection.

> so this feels like we missed types like TagsMap

What is `TagsMap`? Tags are not a map, it's an array (list). There is `Tag_List`.

context | [Context object](#context-object) | The context influences evaluation for assertions. With `matchParentKeys` or `excludeParentKeys` it evaluates the subset of subject type. The resolution of reference objects is done at the context level. If no context is provided, it evaluates the assertion for all instances of the given type. See [context example](#context-example). 
message | `string` | Problem message displayed if the assertion is false.
suggest | [`string`] | List of suggestions to display if the problem occurs.
severity | `string` | The severity level of the problem if the assertion is false. It must be one of these values: `error`, `warn`, `off`. Default value is `error`.
enum | [`string`] | Asserts value is within a predefined list of values. See [enum example](#enum-example).
pattern | `string` | Asserts value matches a regex pattern. See [regex pattern example](#pattern-example).
casing | `string` | Asserts a casing style from this possible list: `camelCase`, `kebab-case`, `snake_case`, `PascalCase`, `MACRO_CASE`, `COBOL-CASE`, `flatcase`. See [casing example](#casing-example).

> mutuallyExclusive | [`string`] | Asserts list of properties (key names only) are mutually exclusive. Use `$keys` in the `on` when this is enabled (for example, `Operation.$keys`). See [mutuallyExclusive example](#mutuallyexclusive-example).

Reviewer 3: ```suggestion
mutuallyExclusive | [`string`] | Asserts list of properties (key names only) are mutually exclusive. See [mutuallyExclusive example](#mutuallyexclusive-example).
```

> mutuallyRequired | [`string`] | Asserts list of properties (key names only) are mutually required. Use `$keys` in the `on` when this is enabled (for example, `Operation.$keys`). See [mutuallyRequired example](#mutuallyrequired-example).

Reviewer 3: ```suggestion
mutuallyRequired | [`string`] | Asserts list of properties (key names only) are mutually required. See [mutuallyRequired example](#mutuallyrequired-example).
```

      - subject: PathItem
        message: Every path item must have a GET operation.
        required: 
          - get
```

The following example asserts that Tags have both name and description defined.

```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        property:
          - name
          - description
        message: Every tag must have a name and description.
        defined: true
```

Another way to compose that rule using the subject's keys.

```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        message: Every tag must have a name and description.
        required:
          - name
          - description
```

### `context` example

The following example asserts that PUT responses with HTTP status 200 or 201 cannot return an application/pdf content type.
Without the `context`, the assertion would evaluate every MediaTypeMap including:
- Responses with all codes including others than 200 or 201
- Responses for all HTTP methods including delete, get, post, and more.
To restrict the evaluation, use the context feature to limit where context is used.

```yaml
- context:
    - type: Operation
      matchParentKeys: [put]
    - type: ResponsesMap

>       matchParentKeys: [201, 200]

Reviewer 3: ```suggestion
      matchParentKeys: ['201', '200']
```

        severity: error
        pattern: /test/
```

### `casing` example

The following example asserts the casing style is `PascalCase` for named Examples map keys.

```yaml
lint:
  rules:
    assertions:
      - subject: NamedExamples
        message: NamedExamples key must be in PascalCase
        severity: error
        casing: PascalCase
```

Casing supports these styles: 
- camelCase
- COBOL-CASE
- flatcase
- kebab-case
- snake_case
- PascalCase
- MACRO_CASE

### `mutuallyExclusive` example

The following example asserts the operation `description` and `externalDocs` must be mutually exclusive.
This assert runs only on node's keys.

```yaml
lint:
  rules:
    assertions:
      - subject: Operation
        message: "Operation must not define both properties together: description and externalDocs"
        severity: error
        mutuallyExclusive:
          - description
          - externalDocs
```

### `mutuallyRequired` example

The following example asserts that a response body schema must have both `amount` and `currency` properties (and not either one by itself).
This assertion runs only on node's keys.

```yaml Schema example
lint:
  rules:
    assertions:

>       - subject: Response

Reviewer 3: ```suggestion
      - subject: SchemaProperties
         context:
           - type: Response
```

        message: The created_at and updated_at properties are mutually required
        severity: error
        mutuallyRequired:
          - created_at
          - updated_at
```

The following example asserts that when `PUT` requests have either `200` or `201` defined that both `200` and `201` responses defined.

```yaml Response example
lint:
  rules:
    assertions:

>       - subject: ResponseMap

Reviewer 3: ```suggestion
      - subject: ResponsesMap
```

        context:
          - type: Operation
            matchParentKeys:
              - put
        message: Must mutually define 200 and 201 responses for PUT requests.
        severity: error
        mutuallyRequired:
          - '200'
          - '201'
```

### `required` example

The following example asserts that `PUT` requests have both `200` and `201` responses defined.
The difference between `mutuallyRequired` is that neither `200` and `201` need to be defined for it to meet `mutuallyRequired` evaluations.

```yaml Response example
lint:
  rules:
    assertions:
      - subject: ResponseMap
        context:
          - type: Operation
            matchParentKeys:
              - put
        message: Must define 200 and 201 responses for PUT requests.
        severity: error
        required:
          - '200'
          - '201'
```

### `disallowed` example

The following example asserts that `x-code-samples` and `x-internal` are not defined.

```yaml
lint:
  rules:
    assertions:
      - subject: Operation

>         message: x-codeSamples and x-internal must not be defined

Reviewer 3: ```suggestion
        message: x-code-samples and x-internal must not be defined
```
